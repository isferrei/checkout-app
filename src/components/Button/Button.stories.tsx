import type { Meta, StoryObj } from "@storybook/react";

import { Button } from "./Button";

// More on how to set up stories at: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta = {
  title: "Example/Button",
  component: Button,
  tags: ["autodocs"],
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen",
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof Button>;

export const Default = {
  args: {
    type: "button",
    variation: "primary",
    children: "Lorem Ipsum",
  },
} satisfies Story;
