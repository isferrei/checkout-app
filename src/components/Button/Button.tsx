/* eslint-disable react/button-has-type */
import classNames from "classnames";
import type { ButtonHTMLAttributes, ReactNode } from "react";

type ButtonTheme = "primary" | "secondary";

type ButtonProps = {
  variation: ButtonTheme;
  height?: number;
  weight?: number;
  type?: string;
};

type Props = ButtonHTMLAttributes<HTMLButtonElement>;

const Button = (props: ButtonProps & Props) => {
  const handleThemes = (buttonTheme: ButtonTheme) => {
    switch (buttonTheme) {
      case "primary":
        return "bg-blue text-white disabled:bg-gray hover:bg-lightBlue transition-all";
      case "secondary":
        return classNames("text-blue");
    }
  };

  return (
    <button
      type={props.type ?? "button"}
      style={{
        height: props.height ? `${props.height}rem` : "50px",
        fontWeight: props.weight ? "700" : "400",
      }}
      className={classNames(
        "flex justify-center items-center w-full rounded-[25px] text-[0.94rem]",
        handleThemes(props.variation)
      )}
    >
      {props.children}
    </button>
  );
};

export { Button };
