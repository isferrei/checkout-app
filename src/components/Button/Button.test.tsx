import { render, screen } from "@testing-library/react";
import type { ReactNode } from "react";

import { Button } from "./Button";

// Mock `next/head`: https://bradgarropy.com/blog/mocking-nextjs
jest.mock(
  "next/head",
  () =>
    function Head(props: { children: ReactNode }) {
      // eslint-disable-next-line testing-library/no-node-access
      return props.children;
    }
);

describe("Button component", () => {
  describe("Render method", () => {
    it("should have label, variation, and type", async () => {
      render(<Button variation="primary">Procurar</Button>);

      const button = screen.getByText(/Procurar/);

      expect(button).toHaveTextContent("Procurar");
      expect(button).toHaveClass("bg-blue");
    });

    it("should have correct class for secondary variation", () => {
      render(<Button variation="secondary">Secondary Button</Button>);

      const button = screen.getByText(/Secondary Button/);

      expect(button).toHaveTextContent("Secondary Button");
      expect(button).toHaveClass("text-blue");
    });

    it("should have custom height and fontWeight styles", () => {
      render(
        <Button variation="primary" type="button" height={3} weight={700}>
          Custom Button
        </Button>
      );

      const button = screen.getByText(/Custom Button/);

      expect(button).toHaveStyle({ height: "3rem", fontWeight: "700" });
    });
  });
});
