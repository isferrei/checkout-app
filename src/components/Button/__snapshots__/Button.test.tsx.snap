// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Render method should have correct class for secondary variation 1`] = `
<button
  class="flex justify-center items-center w-full rounded-[25px] text-[0.94rem] text-blue"
  style="height: 50px; font-weight: 400;"
  type="button"
>
  Secondary Button
</button>
`;

exports[`Button component Render method should have custom height and fontWeight styles 1`] = `
<button
  class="flex justify-center items-center w-full rounded-[25px] text-[0.94rem] bg-blue text-white disabled:bg-gray hover:bg-lightBlue transition-all"
  style="height: 3rem; font-weight: 700;"
  type="button"
>
  Custom Button
</button>
`;

exports[`Button component Render method should have label, variation, and type 1`] = `
<button
  class="flex justify-center items-center w-full rounded-[25px] text-[0.94rem] bg-blue text-white disabled:bg-gray hover:bg-lightBlue transition-all"
  style="height: 50px; font-weight: 400;"
  type="button"
>
  Procurar
</button>
`;
